doc/tkill.1
.\"	This manual page was written by Jan Schaumann <jschauma@yahoo-inc.com>
.Dd February 02, 2011
.Dt TKILL 1
.Os
.Sh NAME
.Nm tkill
.Nd kill a process if it has been running for too long
.Sh SYNOPSIS
.Nm
.Op Fl Phl
.Op Fl t Ar timeout
.Op Fl s Ar signal
.Ns ( Ns Ar pid Ar ... | Fl C Ar cmd | Fl c Ar cmd Ns )
.Sh DESRIPTION
The
.Nm
utility kills a given process or process group if it has been running for
longer than the specified time.
.Pp
The following options are supported:
.Bl -tag -width t_timeout_
.It Fl C Ar cmd
Kill processes by name, not by process ID.
If the given command is found to be a substring of any command in the
process table, then those processes are killed.
.It Fl P
Kill all processes whose parent is the given pid.
.It Fl c Ar cmd
Kill processes by name, not by process ID.
The given command must match exactly the command in the process table.
.It Fl h
Print a usage statement and exit.
.It Fl l
Print a list of valid signals and exit.
.It Fl t Ar timeout
Specify the timeout in seconds.
If not specified, default to 300.
.It Fl s Ar signal
Kill with the given signal, which may be either numeric or the symbolic
name of hte signal.
If not specified, kill with the TERM signal.
.El
.Sh EXAMPLES
The following examples illustrate common usage of this tool.
.Pp
To kill the processes with the process IDs 1234 1235 and 1236 if they have
been running for longer than 500 seconds:
.Bd -literal -offset indent
tkill 1234 1235 1236
.Ed
.Pp
To kill all the children of process ID 1234 if they have been running for
longer than one hour:
.Bd -literal -offset indent
tkill -t 3600 -P 1234
.Ed
.Sh SEE ALSO
.Xr kill 1
.Sh HISTORY
The
.Nm
utility was originally written by
.An Jan Schaumann
.Aq jschauma@yahoo-inc.com
in July 2007.
.Sh BUGS
Please submit bug reports and feature requests to the author.


CHANGES
Version 3.1.4.1 (2011-02-05)
 * add autopw password injector

Version 3.1.4 (2011-01-29)
 * aggregator scripts should be passed date the scan was
   started
 * don't use possibly PREFIX variable in the generic aggregator -- it is
   set in the scanmaster already

Version 3.1.3 (2011-01-28)
 * remove hardcoded references users, keys, hostnames,
   pathnames etc.

Version 3.1.2 (2011-01-22)
 * ensure all output is collected in aggregator

Version 3.1.1 (2011-01-19)
 * let the aggregator also collect hosts_exit_* files

Version 3.1.0 (2010-12-11)
 * allow use to specify password-only, pubkey-only or
   either-or via -A authmode; allow user to set SSHOPTS in environment
 * allow user to set SSH_ADD_OPTS to influence ssh-add(1) behaviour
 * allow user to set SCAN_ENV to influence all scanslave(1) processes

Version 3.0.0 (2010-11-23)
 * use a headless user by default

Version 2.3.3 (2010-08-23)
 * add generic wrapper scripts

Version 2.3.2 (2010-08-12)
 * if input has too few lines, don't try to split

Version 2.3.1 (2010-08-04)
 * invoke tkill every minute; otherwise a hung job will be terminated only
   after 10 minutes (worst case)
 * try to kill a possibly running ssh-agent when all is said and done
 * tkill all processes running for longer than 120 seconds

Version 2.3.0 (2010-07-21)
 * let scanslave start an ssh-agent;
   change protocol order to 2,1: not all hosts have an sshv1 key, so
   attempting sshv1 first with a key and failing will end the attempts
 * explicitly disable agent forwarding when connecting to the target hosts

Version 2.2.2 (2009-11-07)
 * remove slave status file after it finished

Version 2.2.1 (2009-11-07)
 * fix syntax error introduced in 2.2

Version 2.2 (2009-11-07)
 * add suggestion for a wrapper if aggregator args aren't
   sufficiently flexible
 * offer a way to show what slaves are still working (done
   by way of SIGINFO)
 * correct use of ssh agent forwarding

Version 2.1 (2009-09-14)
 * explicitly set StrictHostKeyChecking=no
 * don't interpret '*' as a filename if not enough hosts
   are given on input to actually create a split
 * don't require optional files/arguments
 * clarify that a number of scripts passed to scanmaster
   are passed through to all scanslaves

Version 2.0 (2009-08-02)
 * use , as OFS

Version 1.1 (2009-08-02)
 * make remote exit codes useful
 * add examples to scanslave and scanmaster manpages
 * let scanmaster pass additional flags to scanslave

Version 1.0 (2009-05-18)
 * run the pre-script before checking for existence of other given files,
   since the pre-script might generate some of them
 * don't invoke an ssh-agent; if users want to use an
   agent, they need to start it on their desktop and forward it
   (change in behaviour, so bumping major)

Version 0.1 (2009-05-07)
 * Initial packaged version


src/tkill.py
#! /usr/bin/env python
#
# Copyright (c) 2007,2011 Yahoo! Inc.
# All rights reserved.
#
# Originally written by Jan Schaumann <jschauma@yahoo-inc.com> in July 2007.
#
# Redistribution and use of this software in source and binary forms,
# with or without modification, are permitted provided that the following
# conditions are met:
#
# * Redistributions of source code must retain the above
#   copyright notice, this list of conditions and the
#   following disclaimer.
#
# * Redistributions in binary form must reproduce the above
#   copyright notice, this list of conditions and the
#   following disclaimer in the documentation and/or other
#   materials provided with the distribution.
#
# * Neither the name of Yahoo! Inc. nor the names of its
#   contributors may be used to endorse or promote products
#   derived from this software without specific prior
#   written permission of Yahoo! Inc.
#
# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS
# IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED
# TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A
# PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
# OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
# SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
# LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
# DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
# THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
# (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
# OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
#
#
# This program kills a given process or process group if it has been running
# for longer than a specified time.
#
# See http://mail-index.netbsd.org/tech-userlevel/2007/07/26/0003.html for
# a patch to NetBSD's pkill(1) sources.

import getopt
import os
import re
import string
import sys
import time

###
### Globals
###

EXIT_ERROR = 1
EXIT_SUCCESS = 0

PS_CMD = "ps -awwx -o lstart,pid,ppid,command"

SIGNALS = [ "HUP",  "INT",  "QUIT", "ILL",  "TRAP", "ABRT", "EMT",  "FPE",
            "KILL", "BUS",  "SEGV", "SYS",  "PIPE", "ALRM", "TERM", "URG",
            "STOP", "TSTP", "CONT", "CHLD", "TTIN", "TTOU", "IO",   "XCPU",
            "XFSZ", "VTALRM","PROF","WINCH","INFO", "USR1", "USR2", "PWR" ]

CMD = ""
LOOSE_MATCH = False
PARENT = False
PIDS = []
SIG = "TERM"
TIMEOUT = 300

###
### Subroutines
###

def usage():
    """print short usage"""

    print 'Usage: %s [-P] [-h] [-l] [-t timeout] [-s signal] (pid [...] | -c cmd)'  \
                % sys.argv[0]
    print '\t-C cmd      kill processes containing the given command string'
    print '\t-P          kill all processes whose parent is pid'
    print '\t-c cmd      kill processes matching given command exactly'
    print '\t-h          print this help and exit'
    print '\t-l          print list of valid signals'
    print '\t-t timeout  specify timeout in seconds (default: 300)'
    print '\t-s signal   kill with given signal (default: TERM)'



def isValidSignal(sig):
    """validate the given signal by checking if it is a valid numerical signal
       or as a string in the array of signals we know"""

    try:
        s = string.atoi(sig)
        if (s < 1) or (s > 32):
            raise ValueError("illegal numerical signal %d" % sig)
    except:
        for s in SIGNALS:
            if sig == s:
                return sig
        raise ValueError("illegal numerical signal %s" % sig)

    return sig



def parseOpts():
    """parse command-line options and act on them"""

    global CMD, LOOSE_MATCH, PARENT, SIG, TIMEOUT

    try:
        opts, args = getopt.getopt(sys.argv[1:], "C:Pc:hlt:s:")
    except getopt.GetoptError:
        usage()
        sys.exit(EXIT_ERROR)
        # NOTREACHED

    for o, a in opts:
        if o in ("-C"):
            CMD = a
            LOOSE_MATCH = True
        if o in ("-P"):
            PARENT = True
        if o in ("-c"):
            CMD = a
        if o in ("-h"):
            usage()
            sys.exit(EXIT_SUCCESS)
            # NOTREACHED
        if o in ("-l"):
            for s in SIGNALS:
                sys.stdout.write("%s " % s)
            sys.stdout.write("\n")
            sys.exit(EXIT_SUCCESS)
            # NOTREACHED
        if o in ("-t"):
            try:
                TIMEOUT = string.atoi(a)
            except:
                sys.stderr.write('%s: -t: not a number: "%s"\n' % (sys.argv[0], a))
                sys.exit(EXIT_ERROR)
                # NOTREACHED
        if o in ("-s"):
            try:
                SIG = isValidSignal(a)
            except ValueError:
                sys.stderr.write('%s: -s: bad signal `%s\'\n' % (sys.argv[0], a))
                sys.exit(EXIT_ERROR)
                # NOTREACHED

    if args:
        if CMD:
            usage()
            sys.exit(EXIT_ERROR)
            # NOTREACHED
        else:
            try:
                for p in args:
                    PIDS.append(string.atoi(p))
            except ValueError:
                sys.stderr.write('%s: bad pid `%s\'\n' % (sys.argv[0], p))
                sys.exit(EXIT_ERROR)
                # NOTREACHED
    else:
        if not CMD:
            usage()
            sys.exit(EXIT_ERROR)
            # NOTREACHED



def isStalePid(start):
    """return true if the given processes running time has been longer than
       the limit"""

    global TIMEOUT

    now = time.time()
    if start < (now - TIMEOUT):
        return True

    return False



def killAllStalePids():
    """open a pipe to the PS command and parse the output"""

    global CMD, LOOSE_MATCH, PARENT, PIDS, PS_CMD

    ps = os.popen(PS_CMD)
    while 1:
        line = ps.readline()
        if not line:
            break
        p = re.compile("^(\S+ \S+ +\d+ [0-9:]+ \d+)\s+(\d+)\s+(\d+)\s+(.*)$")
        m = p.match(line)

        if not m:
            continue

        (start, pid, ppid, comm) = m.groups()

        t = time.strptime(start)
        start = time.mktime(t)

        pid = string.atoi(pid)
        ppid = string.atoi(ppid)

        if ( ((CMD == comm.strip()) or \
             (LOOSE_MATCH and (comm.find(CMD) >= 0))) \
             and isStalePid(start) ):
            killPid(pid)

        for p in PIDS:
            if not isStalePid(start):
                continue
            if PARENT:
                if p == ppid:
                    killPid(p)
            else:
                if p == pid:
                    killPid(p)



def killPid(p):
    """kill a process"""

    s = SIGNALS.index(SIG) + 1
    try:
        os.kill(p,s)
    except Exception, (errno, strerror):
        sys.stderr.write('%s: unable to kill PID %d: %s\n' % \
                            (sys.argv[0], p, strerror))

###
### Main
###

if __name__ == "__main__":
    parseOpts()
    killAllStalePids()
